odl/discr/discr_ops.py

```diff
@@ -0,0 +1,477 @@
+    def __init__(self, space, axis=0, dx=1.0, edge_order=2,
+                 zero_padding=False):
+        """Initialize an operator instance.
+
+        Parameters
+        ----------
+        space : `DiscreteLp`
+            The space of elements which the operator is acting on
+        axis : `int`, optional
+            The axis along which the partial derivative is evaluated.
+            Default: 0
+        dx : `float`, optional
+            Scalars specifying the sampling distances in dimension ``axis``.
+            Default distance: 1.0
+        edge_order : {1, 2}, optional
+            First-order accurate differences can be used at the boundaries
+            if no zero padding is used. Default edge order: 2
+        zero_padding : `bool`, optional
+            Implicit zero padding. Assumes values outside the domain of ``f``
+            to be zero. Default: `False`
+        """
+
+        if not isinstance(space, DiscreteLp):
+            raise TypeError('space {!r} is not a `DiscreteLp` '
+                            'instance.'.format(space))
+
+        super().__init__(domain=space, range=space, linear=True)
+        self.axis = axis
+        self.dx = dx
+        self.edge_order = edge_order
+        self.zero_padding = zero_padding
+
+    def _apply(self, x, out):
+        """Apply gradient operator to ``x`` and store result in ``out``.
+
+        Parameters
+        ----------
+        x : ``domain`` element
+            Input vector to which the operator is applied to
+        out : ``range`` element
+            Output vector to which the result is written
+
+        Examples
+        --------
+        >>> from odl import uniform_discr_fromspace, FunctionSpace, Rectangle
+        >>> data = np.array([[ 0.,  1.,  2.,  3.,  4.],
+        ...                  [ 0.,  2.,  4.,  6.,  8.]])
+        >>> space = FunctionSpace(Rectangle([0, 0], [2, 1]))
+        >>> disc = uniform_discr_fromspace(space, data.shape)
+        >>> par_div = DiscretePartDeriv(disc)
+        >>> f = par_div.domain.element(data)
+        >>> par_div_f = par_div(f)
+        >>> print(par_div_f)
+        [[0.0, 1.0, 2.0, 3.0, 4.0],
+         [0.0, 1.0, 2.0, 3.0, 4.0]]
+        """
+
+        if isinstance(self.domain.dspace, Rn):
+            finite_diff(x.asarray(), out.asarray(), axis=self.axis,
+                        dx=self.dx, edge_order=self.edge_order,
+                        zero_padding=self.zero_padding)
+        elif isinstance(self.domain.dspace, CudaRn):
+            raise NotImplementedError('support for `CudaRn` to be '
+                                      'implemented.')
+        else:
+            raise TypeError('invalid data space {}.'.format(
+                self.domain.dspace))
+
+    @property
+    def adjoint(self):
+        """Return the adjoint."""
+        raise NotImplementedError('adjoint not implemented')
+
+
+class DiscreteGradient(Operator):
+    """Spatial gradient operator for `DiscreteLp` spaces with any number of
```

Comment:
Top doc should fit inside one row:

`Spatial gradient operator for `DiscreteLp` spaces.`

or similar.